File: api/src/main/java/com/example/api/constraint/StartBeforeEnd.java
package com.example.api.constraint;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Constraint(validatedBy = StartBeforeEndValidator.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface StartBeforeEnd {

    String message() default StartBeforeEndValidator.MESSAGE;

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: api/src/main/kotlin/com/example/api/constraint/StartBeforeEnd.kt
package com.example.api.constraint

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Constraint(validatedBy = [StartBeforeEndValidator::class])
@MustBeDocumented
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS)
annotation class StartBeforeEnd(

    val message: String = StartBeforeEndValidator.MESSAGE,

    val groups: Array<KClass<*>> = [],

    val payload: Array<KClass<out Payload>> = []
)
